src/services/api.service.js

```diff
@@ -12,19 +13,162 @@ export function get (endpoint) {
+                    }
+                    resolve(book);
                 }, serverLatency);
             });
     });
 }
+
+export function createBook (book) {
+    return new Promise((resolve, reject) => {
+        getBookByIsbn(book.isbn)
+            .catch(() => {
+                validateBook(book).catch((error) => reject(error));
+
+                const defaultValues = {
+                    pageCount: -1,
+                    publishedDate: '',
+                    thumbnailUrl: 'https://picsum.photos/150/190',
+                    shortDescription: '',
+                    longDescription: '',
+                    categories: [],
+                    favorite: false,
+                };
+                const bookToCreate = {
+                    ...defaultValues,
+                    ...book,
+                };
+
+                fetch(`../data/books.json`, {
+                    method: 'POST',
+                    headers: {
+                        'Content-Type': 'application/json'
+                    },
+                    body: JSON.stringify(bookToCreate),
+                }).then(() => {
+                    setTimeout(() => {
+                        resolve();
+                    }, serverLatency);
+                });
+            })
+            .then((book) => {
+                reject(new Error('Book already exists', book));
+            });
+    });
+}
+
+export function updateBook (isbn, updates) {
+    return new Promise((resolve, reject) => {
+        getBookByIsbn(isbn)
+            .catch((error) => reject(error))
+            .then((currentBook) => {
+                const bookWithUpdates = {
+                    ...currentBook,
+                    ...updates,
+                };
+                validateBook(bookWithUpdates).catch((error) => reject(error));
+
+                fetch(`../data/books.json`, {
+                    method: 'PUT',
+                    headers: {
+                        'Content-Type': 'application/json'
+                    },
+                    body: JSON.stringify(bookWithUpdates),
+                }).then(() => {
+                    setTimeout(() => {
+                        resolve();
+                    }, serverLatency);
+                });
+            });
+    });
+}
+
+export function deleteBook (isbn) {
+    return new Promise((resolve, reject) => {
+        getBookByIsbn(isbn)
+            .catch((error) => reject(error))
+            .then(() => {
+                fetch(`../data/books.json`, {
+                    method: 'DELETE',
+                    headers: {
+                        'Content-Type': 'application/json'
+                    },
+                    body: JSON.stringify({ isbn }),
+                }).then(() => {
+                    setTimeout(() => {
+                        resolve();
+                    }, serverLatency);
+                });
+            });
+    });
+}
+
+function validateBook (book) {
```

Comment:
Maybe delete a couple of these and add them to one of the tasks so they can learn about backen validation?

Reply:
I think again as mentioned before this probably isn't the best representation of backend to learn those concepts. For the workshop we should probably have them ignore it completely and for the workexperience if they get through it quickly it would probably be better for them to implement a backend properly with a database and endpoints etc. rather than messing around with this faked api service. What do you think?

Reply:
Sounds sensible